base/gmp.jl

```diff
@@ -682,8 +682,16 @@ end
 
 factorial(x::BigInt) = isneg(x) ? BigInt(0) : MPZ.fac_ui(x)
 
+function binomial(n::BigInt, k::Integer)
+    n < 0 && return (-1)^k * binomial(k - n - 1, k)
+    k < 0 && return BigInt(0)
+    k <= typemax(UInt) && return binomial(n, UInt(k))
```

Comment:
This makes the comment case where `k` fits into an `UInt` slower:
```
julia> function binomial2(n::BigInt, k::Integer)
            n < 0 && return (-1)^k * binomial(k - n - 1, k)
            k < 0 && return BigInt(0)
            k <= typemax(UInt) && return binomial(n, UInt(k))
            κ = n - k
            κ < 0 && return BigInt(0)
            κ <= typemax(UInt) && return binomial(n, UInt(κ))
            throw(OverflowError("Computation will exceed memory"))
        end
binomial2 (generic function with 1 method)

julia> x=-big(2)^30
-1073741824

julia> @btime binomial(x, 5)
  309.237 ns (10 allocations: 128 bytes)
-11893730883318665685294551092682921806397440

julia> @btime binomial2(x, 5)
  468.750 ns (19 allocations: 288 bytes)
-11893730883318665685294551092682921806397440
```

This can be avoided by using the GMP code even for negative `n` if possible, i.e., just reordering your code a bit
```suggestion
    k < 0 && return BigInt(0)
    k <= typemax(UInt) && return binomial(n, UInt(k))
    n < 0 && return (-1)^k * binomial(k - n - 1, k)
```

Reply:
Also `(-1)^k` is far from efficient. Much better to check `isodd(k)` and change the sign then.

Finally, if you really want to cover as much as possible, you could also catch the case where `k > n` and return 0 then, too.

Reply:
Thanks I have implemented the suggested changes.

Although `isodd(k)` seems only marginally better?
```
julia> @benchmark (-1)^(big(2)^23000+rand(1:2))
BenchmarkTools.Trial: 10000 samples with 162 evaluations.
 Range (min … max):  680.660 ns … 535.353 μs  ┊ GC (min … max):  0.00% … 95.98%
 Time  (median):     855.556 ns               ┊ GC (median):     0.00%
 Time  (mean ± σ):     1.303 μs ±   7.375 μs  ┊ GC (mean ± σ):  18.37% ±  5.63%

  █▃                                                            ▁
  ██▆▄▁▁▁▁▁▁▁▃▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▄▇█ █
  681 ns        Histogram: log(frequency) by time       17.8 μs <

 Memory estimate: 5.82 KiB, allocs estimate: 10.

julia> @benchmark isodd(big(2)^23000+rand(1:2))
BenchmarkTools.Trial: 10000 samples with 178 evaluations.
 Range (min … max):  654.708 ns … 485.639 μs  ┊ GC (min … max):  0.00% … 96.26%
 Time  (median):     813.039 ns               ┊ GC (median):     0.00%
 Time  (mean ± σ):     1.174 μs ±   6.547 μs  ┊ GC (mean ± σ):  16.78% ±  5.68%

  █▆                                                            ▁
  ███▅▄▁▄▃▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▄▇██ █
  655 ns        Histogram: log(frequency) by time       12.9 μs <

 Memory estimate: 5.77 KiB, allocs estimate: 8.

```

> Finally, if you really want to cover as much as possible, you could also catch the case where k > n and return 0 then, too.

```
    κ = n - k
    κ < 0 && return BigInt(0)
```
Should take care of that, I think.